lib/rabbitmq.js

```diff
@@ -0,0 +1,121 @@
+const amqp = require('amqplib')
+
+class RabbitMQ {
+  #url
+  #connection
+  #channel
+  #checkExchangeChannel
+  #logger
+  #generateExchange
+
+  constructor (opts = {}) {
+    const { logger, generateExchange } = opts
+    this.#logger = logger
+    this.#connection = null
+    this.#channel = null
+    this.#checkExchangeChannel = null // separate channel to check if exchange exists because the checkExchange method closes the channel
+    this.#generateExchange = generateExchange
+    console.log('RabbitMQ', this.#generateExchange)
+  }
+
+  async connect (url) {
+    const logger = this.#logger
+    if (this.#connection) {
+      this.#logger.warn('Connection already established, reopening')
+      this.#connection.close()
+      this.#connection = null
+      this.#channel = null
+      this.#checkExchangeChannel = null
+    }
+    try {
+      this.#connection = await amqp.connect(url)
+    } catch (err) {
+      throw new Error(`Connection failed to ${url}`)
+    }
+    logger.info('Connected to RabbitMQ', url)
+  }
+
+  async #getChannel () {
+    const logger = this.#logger
+    if (!this.#connection) {
+      logger.warn('Connection not established')
+      await this.connect()
+    }
+
+    if (this.#channel) {
+      return this.#channel
+    }
+
+    const channel = await this.#connection.createChannel()
+    this.#channel = channel
+    return channel
+  }
+
+  // If `#generateExchange` is true, assert the exchanges as fanout/not durable, otherwise check if the exchange exists
+  // and throws if not.
+  async #checkExchange (exchange) {
+    if (this.#generateExchange) {
+      const channel = await this.#getChannel()
+      await channel.assertExchange(exchange, 'fanout', { durable: false })
+      return
+    }
+
+    const logger = this.#logger
+    if (!this.#checkExchangeChannel) {
+      this.#checkExchangeChannel = await this.#connection.createChannel()
+    }
+    this.#checkExchangeChannel.on('error', (err) => {
+      logger.error('Failed to check exchange', err)
+    })
+    try {
+      await this.#checkExchangeChannel.checkExchange(exchange)
+    } catch (err) {
+      logger.error('The exchange does not exists', exchange)
+      this.#checkExchangeChannel = null // the channel is now invalid
+      if (err.code === 404) {
+        throw new Error(`Exchange ${exchange} does not exist`, err)
+      }
+      throw err
+    }
+  }
+
+  async listen (exchange, routingKey, consumer) {
+    const logger = this.#logger
+    const channel = await this.#getChannel()
+    await this.#checkExchange(exchange)
+    const { queue } = await channel.assertQueue('', { exclusive: true })
```

Comment:
the queue name and exclusive should be configurable. Also `durable` should be there.